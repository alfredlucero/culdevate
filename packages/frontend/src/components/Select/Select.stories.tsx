import React from "react";
import { storiesOf } from "@storybook/react";
import Select, { SelectOption } from "./index";

const selectOptions: SelectOption[] = [
  { label: "Corgi", value: "corgi" },
  { label: "Dachsund", value: "dachsund" },
  { label: "Golden Retriever", value: "golden retriever" },
];

storiesOf("Common/Select", module)
  .add("Option Selected", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <Select id="select" value="corgi" options={selectOptions} onChange={() => {}} />
    </div>
  ))
  .add("Required Label", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <Select id="select" value="corgi" options={selectOptions} onChange={() => {}} label="Dogs" required={true} />
    </div>
  ))
  .add("Disabled Placeholder", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <Select
        id="select"
        value=""
        options={selectOptions}
        placeholder="Select a dog..."
        onChange={() => {}}
        label="Dogs"
        disabled={true}
      />
    </div>
  ))
  .add("Inline Text Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        label="Dogs"
        valid={true}
        textInfo="Dog Info"
      />
    </div>
  ))
  .add("Inline Error Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <Select
        id="select"
        value=""
        options={selectOptions}
        onChange={() => {}}
        label="Dogs"
        valid={false}
        required={true}
        errorInfo="Please select a dog."
      />
    </div>
  ));
