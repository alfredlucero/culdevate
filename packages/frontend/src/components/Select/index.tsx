import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";

export interface SelectProps extends CommonProps {
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  onBlur?: (e: React.FocusEvent<HTMLSelectElement>) => void;
  value: string;
  options: SelectOption[];
  id: string;
  required?: boolean;
  valid?: boolean;
  errorInfo?: string;
  textInfo?: string;
  label?: string;
  disabled?: boolean;
  placeholder?: string;
}

export interface SelectOption {
  value: string;
  label: string;
}

const selectBaseClasses = [
  "block",
  "appearance-none",
  "w-full",
  "bg-white",
  "border",
  "border-gray-400",
  "hover:border-gray-500",
  "px-4",
  "py-2",
  "pr-8",
  "rounded",
  "shadow",
  "leading-tight",
  "focus:outline-none",
  "focus:shadow-outline",
];
const selectErrorClasses = ["border-red-500"];
const selectDisabledClasses = ["bg-gray-200", "opacity-75"];

const Select: React.FC<SelectProps> = ({
  onChange,
  value,
  options,
  id,
  onBlur = () => {},
  required = false,
  valid = true,
  errorInfo = "",
  textInfo = "",
  label = "",
  disabled = false,
  placeholder = "",
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <div {...(className !== "" ? { className } : {})}>
      {label && (
        <label className="block text-gray-700 text-sm font-bold mb-2" htmlFor={id}>
          {label} {required ? "" : "(optional)"}
        </label>
      )}

      <div className="inline-block relative">
        <select
          className={cn(selectBaseClasses, !valid ? selectErrorClasses : [], disabled ? selectDisabledClasses : [])}
          id={id}
          name={id}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
          disabled={disabled}
          required={required}
          aria-required={required}
          data-testid={testId}
          {...passThroughProps}
        >
          {placeholder && <option value="">{placeholder}</option>}
          {options.map((option, idx) => (
            <option value={option.value} key={idx}>
              {option.label}
            </option>
          ))}
        </select>

        <div className="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700">
          <svg className="fill-current h-4 w-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
            <path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z" />
          </svg>
        </div>
      </div>

      {valid && textInfo !== "" && <p className="text-gray-700 text-xs italic mt-2">{textInfo}</p>}

      {!valid && errorInfo !== "" && <p className="text-red-500 text-xs italic mt-2">{errorInfo}</p>}
    </div>
  );
};

export default Select;
