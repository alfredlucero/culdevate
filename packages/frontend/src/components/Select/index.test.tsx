import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Select, { SelectOption } from "./index";

describe("<Select />", () => {
  const selectTestId = "selectTestId";
  const selectOptions: SelectOption[] = [
    { label: "Corgi", value: "corgi" },
    { label: "Dachsund", value: "dachsund" },
    { label: "Golden Retriever", value: "golden retriever" },
  ];

  test("should render without error", () => {
    const { container } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render label if present", () => {
    const label = "Select Label";
    const { getByLabelText } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
        label={label}
        required={true}
      />,
    );

    expect(getByLabelText(label)).toBeVisible();
  });

  test("should render label with optional if present and not required", () => {
    const label = "Select Label";
    const { getByLabelText, getByTestId } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
        label={label}
        required={false}
      />,
    );
    expect(getByLabelText(`${label} (optional)`)).toBeVisible();
    expect(getByTestId(selectTestId)).not.toBeRequired();
  });

  test("should render inline text info if valid and info present", () => {
    const textInfo = "Some text info";
    const { container } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
        textInfo={textInfo}
        valid={true}
        errorInfo="Some error info not displayed"
        required={false}
      />,
    );

    expect(container).toHaveTextContent(textInfo);
  });

  test("should render inline error info if invalid and error present", () => {
    const errorInfo = "Error info";
    const { container } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
        textInfo="Text info not displayed"
        valid={false}
        errorInfo={errorInfo}
        required={false}
      />,
    );

    expect(container).toHaveTextContent(errorInfo);
  });

  test("should render disabled state if disabled", () => {
    const { getByTestId } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={() => {}}
        testId={selectTestId}
        className="extra-select-class"
        disabled={true}
        required={false}
      />,
    );

    expect(getByTestId(selectTestId)).toBeDisabled();
  });

  test("should call onChange callback when firing change event", () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <Select
        id="select"
        value="corgi"
        options={selectOptions}
        onChange={onChangeMock}
        testId={selectTestId}
        className="extra-select-class"
        required={false}
      />,
    );

    fireEvent.change(getByTestId(selectTestId), { target: { value: "dachsund" } });

    expect(onChangeMock).toHaveBeenCalled();
  });
});
