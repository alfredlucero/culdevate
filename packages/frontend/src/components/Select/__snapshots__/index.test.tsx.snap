// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render without error 1`] = `
<div>
  <div
    class="extra-select-class"
  >
    
    <div
      class="inline-block relative"
    >
      <select
        aria-required="false"
        class="block appearance-none w-full bg-white border border-gray-400 hover:border-gray-500 px-4 py-2 pr-8 rounded shadow leading-tight focus:outline-none focus:shadow-outline"
        data-testid="selectTestId"
        id="select"
        name="select"
      >
        
        <option
          value="corgi"
        >
          Corgi
        </option>
        <option
          value="dachsund"
        >
          Dachsund
        </option>
        <option
          value="golden retriever"
        >
          Golden Retriever
        </option>
      </select>
      <div
        class="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700"
      >
        <svg
          class="fill-current h-4 w-4"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
